build.gradle.kts

```diff
@@ -6,6 +6,10 @@ plugins {
     id("dokkabuild.base")
 }
 
+tasks.check {
```

Comment:
AFAIK we already have this dependency below: `addDependencyOnSameTasksOfIncludedBuilds`

Reply:
Also, in this root build script we have `integrationTest` aggregate task, which, as far as I understand, will now run only maven and cli tests and not Gradle. Are we planning to migrate `maven/cli` tests to test suites soon? If no, I think it will be better to somehow align this.

Reply:
Note: before, when running `check` (`gradle check`) integration tests were not executed, but now they are. It's not convenient as integration tests are long :)

Reply:
> AFAIK we already have this dependency below: `addDependencyOnSameTasksOfIncludedBuilds`


Unfortunately `addDependencyOnSameTasksOfIncludedBuilds` adds a dependency on `:dokka-integration-tests:check`, not `:dokka-integration-tests:gradle:check`.

TBH I found the logic about setting up task dependencies a bit confusing in this file, so there's probably a better way of using the functions, but I found being explicit easier to understand :). 

I get that it's a workaround for a Gradle flaw https://github.com/gradle/gradle/issues/22335 - I think using a task rule might be help https://stackoverflow.com/a/69179714/4161471


> Are we planning to migrate maven/cli tests to test suites soon? 

[Yes, for sure](https://github.com/Kotlin/dokka/pull/3427#discussion_r1432756930). I wanted to do it one subproject at a time, to minimize PR size, and then it will be easier to determine what functionality can be extracted into conventions & build-logic utils.

> Note: before, when running check (gradle check) integration tests were not executed, but now they are. It's not convenient as integration tests are long :)

Yes, that's intentional :) 'check' is the defacto standard lifecycle task for _all_ verification.

Reply:
Let's start from the end:)

> Yes, that's intentional :) 'check' is the defacto standard lifecycle task for all verification.

1. Yes, but in day-to-day workflow I would like to run just `test` + `apiCheck` (so old `check`) to test if my changes didn't broke anything. Sometimes it's fine to run only one module/test, but not all the time. At my side running `check` now will take 1-2 minute I think. If we will put running integration tests into `check` - it will become much more, so I will start to use `test` and `apiCheck` tasks separately. Or I will just run `check` for some subproject(s). External contributors also could just run `check` (as it's `defacto standard lifecycle task for all verification.`) and Im not sure, that all of them will want to wait for all tests or some of the tests could even fail, f.e. android, if there is no setup, and they will spend more time to fix the issue or will just ignore running test, or even finishing the feature. This is a little inconvenient.
3. as far as I understand, test suites created by Gradle test-suite plugin doesn't run when running `check` - and I think for a reason. `check` is day-to-day task for **fast** versification - integration tests are not fast.

So I think, that running `check` should not run integration tests - they should run by default on CI (in same/separate workflow), it should be enough for most cases. If needed, they can be easily run locally, all of them (rarely) or specific one (more frequently).

> I wanted to do it one subproject at a time, to minimize PR size

Agree, though, I've just mentioned inconsistency in behaviour of different integration tests after merging this PR - not a blocker, but just something to be aware of.

> TBH I found the logic about setting up task dependencies a bit confusing in this file, so there's probably a better way of using the functions, but I found being explicit easier to understand :).

yeah, still, it's very convenient to have such `aggregated` tasks in root project. If you have ideas on how to improve this, I will be glad to review separate PR :)
I've never used task rules before, so I don't know will it be possible/simpler/complex, so up to you.

Reply:
Sorry, but I strongly believe that `check` or `test` or any other similar basic task should not run anything as heavyweight / complicated as all of Dokka's "integration tests". 

I don't know if the reworked tests have the same problems as the current ones, but now you are unable to run integration tests locally if you
* didn't initialize git submodules (not done by default, fails with obscure errors like "unable to find a file")
* don't have Android SDK installed (the android test is executed regardless)
* have less than 32gb of ram
* can't leave some hungry process running for 1h (I had to close my IDEs and chrome to run tests when I had 16gb of ram, and it still didn't work half the time)

If you do manage to run these tests and they fail, it's not obvious at all to external contributors or members without context as to what to do about it and how to fix it (especially if it's a false positive or is expected), so for them it's virtually useless.

Integration tests were run by default before, and it was very annoying for the reasons outlined by Oleg. I ran pretty much every build with `test -x integrationTest`, but it might be impossible after this change because there's no taskgraph between tasks of included builds (all the `addDependencyOnSameTasksOfIncludedBuilds` calls), so they will be run whether you want it or not

___

As a side note (not a requirement or a comment for this PR specifically, there's no need to implement it). 

Ideally, I think the workflow should be the following: 
1. run basic checks locally (<= 1-2 min)
2. push a commit that triggers more complex CI checks (smoke, <= 30m), merge the PR
3. even more complex CI checks are triggered for every commit in master / nightly / manually (>= 30m)

 This idea was mentioned in https://github.com/Kotlin/dokka/issues/2925#issuecomment-1840852198:
> The current integration tests take 1h+ to run, which is painful in PRs and hogs CI agents. There should be a a differentiation between "full tests" that run everything, and their subset in the form of "smoke tests", that only test specific versions (maybe only the latest) or projects (maybe only 2 out of 10 user projects). Smoke tests are intended to be run for commits in PRs, and full tests - overnight or triggered manually.

Reply:
Thanks for explaining, it definitely makes sense to make the UX of the Gradle tasks more comprehensible. Anyone should be able to run suitable tests with a single Gradle task.

I see there are some good options for this, proposed at the bottom, but first I want to clarify about the `check` task.

### Check task

Personally, I find the current behaviour (`check` doesn't run all tests) confusing because in all other projects it's a lifecycle task that runs all verifications.

That's because the `check` task is provided by the [Gradle Base Plugin](https://docs.gradle.org/current/userguide/base_plugin.html), which all other plugins use, and it's supposed to run all verifications.

<img width="202" alt="image" src="https://github.com/Kotlin/dokka/assets/152864218/334a5ee3-b30d-44ad-b142-ed32d0ded732">

Of course, Gradle is very flexible so Dokka doesn't have to follow this convention, but it would be unusual. It's important to align the conventions within Kotlin projects - so while "check runs all tests" might not meet your understanding at the moment, I hope it will make more sense in the future.

### Proposal

I see two options

1. Update the `test` task so that it will also run the `apiCheck` task.
    
    Update the README instructions:
    
    * `./gradlew check` - run all verifications (which might be slow):
      * `./gradlew test` - run unit tests (and also `apiCheck`)
      * `./gradlew apiCheck` - BCV check
      * `./gradlew integrationTest` - run all integration tests (Slow! Requires Prep!)

2. Add a custom `quickCheck` (or `checkFast` - whatever name you want) lifecycle task that will run unit tests + apiCheck (and any other future 'fast' checks).

    Update the README instructions:
    
    * `./gradlew check` - run all verification (which might be slow):
      * `./gradlew quickCheck` - runs quicker checks (`test` and `apiCheck`)
      * `./gradlew test` - run unit tests
      * `./gradlew apiCheck` - BCV check
      * `./gradlew integrationTest` - run all integration tests (Slow! Requires Prep!)

I have a slight preference towards 1, because currently there's not many verification tasks apart from `test` and `apiCheck` that would justify a new task.

Which do you prefer?

I can also split out these changes to the test-task dependencies into a separate PR to keep this PR focused.

Reply:
> it's supposed to run all verifications.

Can you try to define "all" here? :) 

I think it's very important to draw the line somewhere, otherwise it's easy to fall into unproductive malicious compliance and make the `check` task perform **literally** all verifications: 
* Running tests on different platforms and not only on the user's one (i.e start docker containers for missing OSes like MacOS/Linux/Windows)
* Publishing verification checks (not done now)
* UI tests that might reside in a different repo (we'll have them soon, #3469)
* Run first with K1, then with K2 (even if not needed by default)
* Tests that run Dokka with the latest/arbitrary commit of `kotlin-compiler` / KGP / etc (also planned)
* Stress tests / Performance tests / Benchmarks

I would draw the line somewhere close to "Runs all embedded verifications that don't require special knowledge or setup, and that are reasonably expected to be run locally by developers". By that definition, Dokka's "integration tests" (which are really completely separate e2e tests) are not it for several reasons

___

About the proposals - I would first try to understand why we need the `check` task to run Dokka's "integration tests" specifically. It's taken for granted, but I don't think it should be :) 

Who or what is going to be using it **if** `check` also runs Dokka's "integration tests"? How is it going to be useful?

Reply:
Yeah, Im totally agree with Ignat.
Still I will add some additional points:
1. AGP doesn't make `check` dependsOn instrumented tests (which are executed on emulator/real device) - they are slow and environment dependent
2. even Gradle by default doesn't do this for test suites - this means, that it's not strictly true, that we need to run all kind of tests by default when we run `check`
3. If we will make `test` dependOn `apiCheck` or any other check task (f.e. if we will add `detekt`) - if we will run tests from IDEA via gutter (just one test or test class), we will execute `apiCheck` - not sure it's expected. May be it's possible to workaround it, but not sure that by doing this we will simplify both build setup and user expectations.

So I believe `integrationTests` for Dokka should be executed in separate task. Additionally, we could created lifecycle task `checkAll`/`checkLong` if we really do want to execute it.

Another possibility: make global `check` run all tests except integration tests. make `dokka-integration-tests` `check` run all tests there, even integration.
It can cause confusion, in this case we will have some split between `check` tasks. But IMO, it will be still better then running all `integration tests` by calling global `check`

Reply:
I've removed the check task dependency.


> > it's supposed to run all verifications.
> 
> Can you try to define "all" here? :)

See https://docs.gradle.org/current/userguide/base_plugin.html

> Plugins and build authors should attach their verification tasks, such as ones that run tests, to this lifecycle task using `check.dependsOn(task)`.



>     * Running tests on different platforms and not only on the user's one (i.e start docker containers for missing OSes like MacOS/Linux/Windows)
>     * Publishing verification checks (not done now)

There are no Gradle tasks for this, so `check` can't depend on them.

>     * UI tests that might reside in a different repo (we'll have them soon, [Infrastructure to run UI tests #3469](https://github.com/Kotlin/dokka/issues/3469))

`check` can't depend on tasks in other repos.

>     * Run first with K1, then with K2 (even if not needed by default)

This is a good idea. It would help prevent the problem we saw recently where the K2 tests weren't running on TeamCity.

>     * Tests that run Dokka with the latest/arbitrary commit of `kotlin-compiler` / KGP / etc (also planned)

If this is achieved with a Gradle task then yes, `check` should depend on that task.

>     * Stress tests / Performance tests / Benchmarks

This depends:

* If the task has checks, then yes, `check` should depend on it

  For example, the existing `SequentialTasksExecutionStressTest`

* If the tasks are measuring performance and reporting on it, but not verifying it, then that's not a verification task so `check` shouldn't depend on it.

   For example: if JMH tests were added


> I would draw the line somewhere close to "Runs all embedded verifications that don't require special knowledge or setup, and that are reasonably expected to be run locally by developers". By that definition, Dokka's "integration tests" (which are really completely separate e2e tests) are not it for several reasons

I agree, there's a good need for such lifecycle task, and the meaning you explain is clear.

It's confusing to overload an existing task with a different meaning. It would make more sense to create a new lifecycle task that meets this definition.

It would also help prevent issues on CI where verifications have been skipped. E.g. K2 tests not being run, and up until recently apiCheck wasn't run on TeamCity.

Using `check` to run all test tasks aligns how `check` works both within JetBrains, but also within the Gradle ecosystem. Many other plugins use `check` to run all verifications.